Add Cyclic Group Representation (CGR) class

Zeldax64 opened the pull request:
<!--- Provide a short summary of your changes in the Title above -->
Introduce a new VSA class. Solves #181.

## Description
<!--- Describe your changes in detail -->
<!-- Link the issue (if any) that will be resolved by the changes -->

CGR is very similar to MCR, but differs in bundling. This pull request introduces:
- A novel BaseMCRTensor parent class, which is used by MCRTensor and CGRTensor.
- CGRTensor, which uses a different bundling than MCRTensor.
- Tests for CGR. It requires just a custom bundling test.

Additionally, this commit also solves a bug in the `MCR: __torch_function__()` implementation. This function parses `args` to search for MCRTensors and ensures they all have the same `block_sizes`.
https://github.com/hyperdimensional-computing/torchhd/blob/6b7478d75ae35197e3e0cdde27a83f410db19965/torchhd/tensors/mcr.py#L420-L424

However, it treats `args` as a shallow 1D container. In this case, the following snippet breaks the code:
```python3
import torch
from torchhd import embeddings

id = embeddings.Random(4, 10, vsa='MCR', block_size=4).weight
t = torch.stack((id[0], id[1]))
```

In this minimal example, `torch.stack()` calls `__torch_function__` with a 2D tuple that can't be parsed.

## Checklist
- [x] I added/updated documentation for the changes.
- [X] I have thoroughly tested the changes.

The only problem at the moment is in the documentation. Due to the addition of `BaseMCRTensor`, the documentation of MCR and CGR is split, requiring users to read two mnual pages. Also, I wasn't able to add new `CGRTensor` and `BaseMCRTensor` classes into the proper table of contents.

Zeldax64 committed:
Add BaseMCRTensor

Add a new base class to support future implementation of variations of
MCR proposals.

Zeldax64 committed:
Fix bug in BaseMCRTensor:__torch_function__()

There is a bug when args is a collection of collections instead of a
plain tuple. In this case, the old args parsing was unable to search for
block_size in nested structures containing BaseMCRTensor.

For instance, let `a` and `b` be BaseMCRTensor variables. Calling
`torch.stack((a, b))` results in error, as the `args` received in
`__torch_function__()` is a nested tuple.

Zeldax64 committed:
Add CGRTensor

Add a new VSA class named Cyclic Group Representation (CGR). This class
is similar to MCR, but differs in bundling.

Zeldax64 committed:
Add CGRTensor to functional embeddings

Allow its usage in level and circular embeddings as done with MCRTensor.

Zeldax64 committed:
Fix CGR bundle

Ensure both inputs are in the same shape.

Zeldax64 committed:
Add CGR into tests

The CGR should behave almost the same as MCR, but diverges in bundling.
Implement a custom bundling test for it.

mikeheddes reviewed (commented):
Thank you for opening this PR, this would be a great addition to the library. Your implementation looks good, I just added some minor comments. Good catch of the `__torch_function__` bug btw.

Zeldax64 committed:
Use Tensor.dim() in CGRTensor

Shorter syntax and more readable.

Zeldax64 committed:
Remove asserts in CGR/MCR classes

No need to check block size in CGR/MCR functions as the
__torch_function__() in BaseMCRTensor already checks it.

Zeldax64 committed:
Improve exception in BaseMCRTensor if multiple block_sizes

- Raise ValueError instead of RuntimeError

Zeldax64 committed:
Add CGRTensor to docs

Include CGRTensor in the README and in the built docs.

Zeldax64 committed:
Add BaseMCRTensor to docs

Zeldax64 commented:
I've just pushed some commits. Regarding your comments,

## Parameter checking

The use of assertions and `RuntimeError` stems from the current MCRTensor implementation. I just copied and pasted the code into BaseMCRTensor. 

> Do `self` and `other` require the same `block_size` for the similarity to make sense? If so, we should add the same error handling as in the `bind` method. However, it might be that the error handling in `__torch_function__` already takes care of this, in which case there is no need for error handling in here or in `bind`

CGR and MCR operations (bind, bundle, and similarity) must always be done with hypervectors adopting the same `block_size`. After reading your comments, I took a deeper look into `__torch_function__` calls, and I believe it is safe to remove any type checking from `bind()` and `bundle()` functions, as `__torch_function__` already excessively checks the parameters. 

In fact, there might be a slight performance issue since `__torch_function__` is called on any PyTorch operation, including functions as  `__get__` or `shape`, which wouldn't require `block_size` checking.

## Documentation

> Then all you have to do to include the `CGRTensor` in the documentation is to add it to docs/torchhd.rst. I would probably go for the second option.

After some tests considering this option, I feel like repeating function signatures also isn't the way to go, as the code isn't DRY. Maybe let users check two doc pages and link `CGR` and `MCR` with `BaseMCRTensor`?

Zeldax64 commented on torchhd/tensors/basemcr.py:
```diff
@@ -0,0 +1,392 @@
+
+        if len(block_sizes) != 1:
+            raise RuntimeError(
+                f"Call to {func} must contain exactly one block size, got {list(block_sizes)}"
```
I believe you misread it for `len(block_sizes)` : )
I don't know why it was converting the dictionary to a list, but either approach would print the same elements.

mikeheddes replied:
You are right, my bad.

Zeldax64 commented on torchhd/tensors/basemcr.py:
```diff
@@ -0,0 +1,392 @@
+        block_sizes = _parse_container_for_attr(args, "block_size")
+
+        if len(block_sizes) != 1:
+            raise RuntimeError(
```
Done

Zeldax64 commented on torchhd/tensors/basemcr.py:
```diff
@@ -0,0 +1,392 @@
+        """
+        return self.clone()
+
+    def dot_similarity(self, others: "BaseMCRTensor", *, dtype=None) -> Tensor:
```
Error handling now in `__torch_function__`.

Zeldax64 committed:
Fix typo in README.md

Zeldax64 committed:
Improve documentation for MCRTensor and CGRTensor

Replicate function signatures in inherited classes (MCRTensor and
CGRTensor) from their parent class (BaseMCRTensor) to let sphinx
print the documentation of each parent function.

mikeheddes reviewed (approved):
Sorry for the late response, I was OOO last week. The test seems indeed to have randomly failed. 

Thank you for the great work. This is a very nice addition to the library.
